compiler/rustc_mir_transform/src/inline.rs

```diff
@@ -388,14 +388,11 @@ impl<'tcx> Inliner<'tcx> {
             return Err("never inline hint");
         }
 
-        // Only inline local functions if they would be eligible for cross-crate
-        // inlining. This is to ensure that the final crate doesn't have MIR that
-        // reference unexported symbols
-        if callsite.callee.def_id().is_local() {
-            let is_generic = callsite.callee.args.non_erasable_generics().next().is_some();
-            if !is_generic && !callee_attrs.requests_inline() {
-                return Err("not exported");
-            }
+        // Reachability pass defines which functions are eligible for inlining. Generally inlining
+        // other functions is incorrect, because they could reference symbols that aren't exported.
```

Comment:
This change is coupling the MIR inliner *even more* to our dubious reachability analysis, and I think that's the wrong way to improve things.

Reply:
MIR inlining needs to agree with the logic used to generate MIR for items and also with whether statics are considered reachable. If it doesn't, we end up needing MIR for a function that we don't have, or we end up requiring visibility for a static that was previously determined to be internal.

This change as written currently is restricting MIR inlining to match the behavior of this function: https://github.com/rust-lang/rust/blob/22d41ae90facbffdef9115809e8b6c1f71ebbf7c/compiler/rustc_passes/src/reachable.rs#L24